### PR TITLE
Improve db_postgres iterators

### PR DESCRIPTION
**Pull request by artemklevtsov**
Reduce memory for the `fastRows` and `InstantRows` iterators with single row mode. Before pgresult consumed memory to allocate the entire result. Now for the single row only.
Docs: https://www.postgresql.org/docs/9.2/libpq-single-row-mode.html

For the testing
Run conteiner:
```bash
docker run --rm --env POSTGRES_HOST_AUTH_METHOD=trust --publish 5432:5432 postgres:13-alpine
```

Code sample:

```nim
let db = open("localhost", "postgres", "", "postgres")
const q = sql "SELECT * FROM pg_catalog.pg_tables LIMIT 3"
for row in db.fastRows(q):
  echo row
for row in db.instantRows(q):
  echo row[1]
var cols: DbColumns
for row in db.instantRows(cols, q):
  echo row[3]
echo cols
echo db.getAllRows(q)
db.close()
```
Result on my local machine:

```bash
@["pg_catalog", "pg_statistic", "postgres", "", "t", "f", "f", "f"]
@["pg_catalog", "pg_type", "postgres", "", "t", "f", "f", "f"]
@["pg_catalog", "pg_foreign_table", "postgres", "", "t", "f", "f", "f"]
pg_statistic
pg_type
pg_foreign_table



@[(name: "schemaname", tableName: "12113", typ: (kind: dbFixedChar, notNull: false, name: "name", size: 0, maxReprLen: 0, precision: 0, scale: 0, min: 0, max: 0, validValues: @[]), primaryKey: false, foreignKey: false), (name: "tablename", tableName: "12113", typ: (kind: dbFixedChar, notNull: false, name: "name", size: 0, maxReprLen: 0, precision: 0, scale: 0, min: 0, max: 0, validValues: @[]), primaryKey: false, foreignKey: false), (name: "tableowner", tableName: "12113", typ: (kind: dbFixedChar, notNull: false, name: "name", size: 0, maxReprLen: 0, precision: 0, scale: 0, min: 0, max: 0, validValues: @[]), primaryKey: false, foreignKey: false), (name: "tablespace", tableName: "12113", typ: (kind: dbFixedChar, notNull: false, name: "name", size: 0, maxReprLen: 0, precision: 0, scale: 0, min: 0, max: 0, validValues: @[]), primaryKey: false, foreignKey: false), (name: "hasindexes", tableName: "12113", typ: (kind: dbBool, notNull: false, name: "bool", size: 0, maxReprLen: 0, precision: 0, scale: 0, min: 0, max: 0, validValues: @[]), primaryKey: false, foreignKey: false), (name: "hasrules", tableName: "12113", typ: (kind: dbBool, notNull: false, name: "bool", size: 0, maxReprLen: 0, precision: 0, scale: 0, min: 0, max: 0, validValues: @[]), primaryKey: false, foreignKey: false), (name: "hastriggers", tableName: "12113", typ: (kind: dbBool, notNull: false, name: "bool", size: 0, maxReprLen: 0, precision: 0, scale: 0, min: 0, max: 0, validValues: @[]), primaryKey: false, foreignKey: false), (name: "rowsecurity", tableName: "12113", typ: (kind: dbBool, notNull: false, name: "bool", size: 0, maxReprLen: 0, precision: 0, scale: 0, min: 0, max: 0, validValues: @[]), primaryKey: false, foreignKey: false)]
@[@["pg_catalog", "pg_statistic", "postgres", "", "t", "f", "f", "f"], @["pg_catalog", "pg_type", "postgres", "", "t", "f", "f", "f"], @["pg_catalog", "pg_foreign_table", "postgres", "", "t", "f", "f", "f"]]
```

### CODE REVIEW

**Comment by ringabout** [comment]
image

The key that signed this is expired. Maybe make a new request?

File: lib/impure/db_postgres.nim [hunk 1]
```diff
@@ -208,44 +223,82 @@ iterator fastRows*(db: DbConn, query: SqlQuery,
...
+      break
+    if pqresultStatus(res) == PGRES_TUPLES_OK:
+      pqclear(res)
+      continue
+    if pqresultStatus(res) != PGRES_SINGLE_TUPLE:
+      dbError(db)
+    var L = pqNfields(res)
+    var result = newRow(L)
+    setRow(res, result, 0, L)
     yield result
```
Comment (Araq) [inline lib/impure/db_postgres.nim]:
```suggestion
    let status = pqresultStatus(res)
    if status == PGRES_TUPLES_OK:
      pqclear(res)
    elif status != PGRES_SINGLE_TUPLE:
      dbError(db)
    else:
      var L = pqNfields(res)
      var result = newRow(L)
      setRow(res, result, 0, L)
      yield result
```

File: lib/impure/db_postgres.nim [hunk 2]
```diff
@@ -208,44 +223,82 @@ iterator fastRows*(db: DbConn, query: SqlQuery,
...
+  var res: PPGresult = nil
+  while true:
+    res = pqgetresult(db)
+    if res == nil:
+      break
+    if pqresultStatus(res) == PGRES_TUPLES_OK:
+      pqclear(res)
+      continue
+    if pqresultStatus(res) != PGRES_SINGLE_TUPLE:
+      dbError(db)
```
Comment (Araq) [inline lib/impure/db_postgres.nim:265]: Same suggestion, use structured programming.

File: lib/impure/db_postgres.nim [hunk 3]
```diff
@@ -208,44 +223,82 @@ iterator fastRows*(db: DbConn, query: SqlQuery,
...
+    yield InstantRow(res: res)
```
Comment (Araq) [inline lib/impure/db_postgres.nim]: Same.

File: lib/impure/db_postgres.nim [hunk 4]
```diff
@@ -208,44 +223,82 @@ iterator fastRows*(db: DbConn, query: SqlQuery,
...
```
Comment (Araq) [inline lib/impure/db_postgres.nim]: Maybe use a template? Copy&Paste programming sucks.

File: lib/impure/db_postgres.nim [hunk 5]
```diff
@@ -382,7 +435,8 @@ proc getColumnType(res: PPGresult, col: int) : DbType =
   of 705:  return DbType(kind: DbTypeKind.dbUnknown, name: "unknown")
   else: return DbType(kind: DbTypeKind.dbUnknown, name: $oid) ## Query the system table pg_type to determine exactly which type is referenced.
 
-proc setColumnInfo(columns: var DbColumns; res: PPGresult; L: int32) =
+proc setColumnInfo(columns: var DbColumns; res: PPGresult) =
```
Comment (Araq) [inline lib/impure/db_postgres.nim]: So the `line` field was useful after all, it saved one call.

File: lib/impure/db_postgres.nim [hunk 6]
```diff
@@ -203,49 +218,72 @@ proc setRow(res: PPGresult, r: var Row, line, cols: int32) =
...
+template fetchRows(db: DbConn): untyped =
+  var res: PPGresult = nil
+  while true:
+    res = pqgetresult(db)
+    if res == nil:
+      break
+    let status = pqresultStatus(res)
+    if status == PGRES_TUPLES_OK:
+      pqclear(res)
+      continue
```
Comment (Araq) [inline lib/impure/db_postgres.nim]:
```suggestion
    if status == PGRES_TUPLES_OK:
      discard
```

File: lib/impure/db_postgres.nim [hunk 7]
```diff
@@ -203,49 +218,72 @@ proc setRow(res: PPGresult, r: var Row, line, cols: int32) =
...
+template fetchinstantRows(db: DbConn): untyped =
+  var res: PPGresult = nil
+  while true:
+    res = pqgetresult(db)
+    if res == nil:
+      break
+    let status = pqresultStatus(res)
+    if status == PGRES_TUPLES_OK:
+      pqclear(res)
+      continue
```
Comment (Araq) [inline lib/impure/db_postgres.nim]:
```suggestion
    if status == PGRES_TUPLES_OK:
      discard
```

File: lib/impure/db_postgres.nim [hunk 8]
```diff
@@ -395,59 +433,80 @@ proc setColumnInfo(columns: var DbColumns; res: PPGresult; L: int32) =
...
+    elif status != PGRES_SINGLE_TUPLE:
+      dbError(db)
+    else:
+      yield InstantRow(res: res)
+  pqclear(res)
```
Comment (Araq) [inline lib/impure/db_postgres.nim]: Outside the loop?

**Pull request merged**